ansible_rulebook/action/run_base.py

```diff
@@ -0,0 +1,224 @@
+            if i > 0:
+                if delay > 0:
+                    await asyncio.sleep(delay)
+                logger.info(
+                    "Previous %s failed. Retry %d of %d",
+                    self._action_name,
+                    i,
+                    retries,
+                )
+
+            retry = await self._do_run()
+            if not retry:
+                break
+
+    async def _pre_process(self) -> None:
+        pass
+
+    async def _post_process(self) -> None:
+        pass
+
+    async def _job_start_event(self):
+        await self.helper.send_status(
+            self._job_data,
+            obj_type="Job",
+        )
+
+
+class RunTemplate(RunBase):
+    """Superclass for template-based run actions.  Launches the appropriate
+    specified template on the controller. It waits for the job to be complete.
+    """
+
+    @property
+    def _exceptions(self) -> tuple:
+        return (ControllerApiException,)
+
+    @property
+    def _job_data(self) -> dict:
+        data = super()._job_data
+        data["hosts"] = ",".join(self.helper.control.hosts)
+        return data
+
+    @property
+    def _run_job(self) -> Callable:
+        raise NotImplementedError
+
+    @property
+    def _template_name(self) -> str:
+        raise NotImplementedError
+
+    @property
+    def _url_path(self) -> str:
+        return self._url_prefix + f"{self.controller_job['id']}/" + "details"
+
+    @property
+    def _url_prefix(self) -> str:
+        return "/#/"
+
+    def _make_log(self) -> dict:
+        log = {
+            "organization": self.organization,
+            "job_id": self.job_id,
+            "status": self.controller_job["status"],
+            "run_at": self.controller_job["created"],
+            "url": self._controller_job_url(),
+            "matching_events": self.helper.get_events(),
+        }
+        if "error" in self.controller_job:
+            log["message"] = self.controller_job["error"]
+            log["reason"] = {"error": self.controller_job["error"]}
+        return log
+
+    def __init__(self, metadata: Metadata, control: Control, **action_args):
+        super().__init__(metadata, control, **action_args)
+        self.organization = self.action_args["organization"]
+        self.job_args = self.action_args.get("job_args", {})
+        self.job_args["limit"] = ",".join(self.helper.control.hosts)
+        self.controller_job = {}
+
+    async def __call__(self):
+        logger.info(
+            "running %s: %s, organization: %s",
+            self._template_name,
+            self.name,
+            self.organization,
+        )
+        logger.info(
+            "ruleset: %s, rule %s",
+            self.helper.metadata.rule_set,
+            self.helper.metadata.rule,
+        )
+
+        self.job_args["extra_vars"] = self.helper.collect_extra_vars(
+            self.job_args.get("extra_vars", {})
+        )
+        await super().__call__()
+
+    async def _do_run(self) -> bool:
```

Comment:
We don't actually need a boolean as result here and I think it is not a good pattern. Error handling should be done through exceptions.

Reply:
We don't have a binary situation; we have success, retriable errors and non-retriable errors.  I find I can live with the idea of using StopAsyncIteration and have pushed a change with that.  If people would rather have a specific exception for the hierarchy let me know.